rust/worker/src/server.rs

```diff
@@ -137,454 +135,230 @@ impl WorkerServer {
         self.system = Some(system);
     }
 
-    pub(crate) async fn query_vectors_instrumented(
+    fn decompose_proto_scan(
         &self,
-        request: Request<QueryVectorsRequest>,
-    ) -> Result<Response<QueryVectorsResponse>, Status> {
-        let request = request.into_inner();
-        let segment_uuid = to_segment_uuid(&request.segment_id)?;
-        let collection_uuid = to_collection_uuid(&request.collection_id)?;
-        let (collection_version, log_position) = get_version_context(&request.version_context)?;
-        let system = self.clone_system()?;
-        let dispatcher = self.clone_dispatcher()?;
-
-        let mut query_vectors = Vec::with_capacity(request.vectors.len());
-        for proto_query_vector in request.vectors {
-            let (query_vector, _encoding) = proto_query_vector
-                .try_into()
-                .map_err(|e| Status::internal(format!("Error converting vector: {}", e)))?;
-
-            query_vectors.push(query_vector);
-        }
-
-        let knn_filter_orchestrator = KnnFilterOrchestrator::new(
-            self.blockfile_provider.clone(),
-            dispatcher.clone(),
-            self.hnsw_index_provider.clone(),
-            // TODO: Load the configuration for this
-            1000,
+        scan: chroma_proto::ScanOperator,
+    ) -> Result<(FetchLogOperator, FetchSegmentOperator), Status> {
+        let collection = scan
+            .collection
+            .ok_or(Status::invalid_argument("Invalid Collection"))?;
+        let collection_uuid = CollectionUuid::from_str(&collection.id)
+            .map_err(|err| Status::invalid_argument(err.to_string()))?;
+        let vector_uuid = SegmentUuid::from_str(&scan.knn_id)
+            .map_err(|err| Status::invalid_argument(err.to_string()))?;
+        let metadata_uuid = SegmentUuid::from_str(&scan.metadata_id)
+            .map_err(|err| Status::invalid_argument(err.to_string()))?;
+        let record_uuid = SegmentUuid::from_str(&scan.record_id)
+            .map_err(|err| Status::invalid_argument(err.to_string()))?;
+        Ok((
             FetchLogOperator {
                 log_client: self.log.clone(),
+                // TODO: Make this configurable
                 batch_size: 100,
-                start_log_offset_id: log_position as u32 + 1,
+                start_log_offset_id: collection.log_position as u32 + 1,
```

Comment:
copy my comment 

// The collection log position is inclusive, and we want to start from the next log
 // Note that we query using the incoming log position this is critical for correctness

Reply:
Added comment